### PR TITLE
Filter identity scopes and offline_access when no explicit scopes are specificed in client credentials

### PR DESCRIPTION
**Pull request by leastprivilege**
see https://github.com/IdentityServer/IdentityServer4/issues/2295

### CODE REVIEW

File: src/IdentityServer4/Validation/TokenRequestValidator.cs [hunk 1]
```diff
@@ -573,21 +573,31 @@ public TokenRequestValidator(IdentityServerOptions options, IAuthorizationCodeSt
...
                 _logger.LogTrace("Client provided no scopes - checking allowed scopes list");
 
                 if (!_validatedRequest.Client.AllowedScopes.IsNullOrEmpty())
                 {
                     var clientAllowedScopes = new List<string>(_validatedRequest.Client.AllowedScopes);
-                    if (_validatedRequest.Client.AllowOfflineAccess)
+                    if (!ignoreImplicitOfflineAccess)
                     {
-                        clientAllowedScopes.Add(IdentityServerConstants.StandardScopes.OfflineAccess);
+                        if (_validatedRequest.Client.AllowOfflineAccess)
```
Comment (LindaLawton) [inline src/IdentityServer4/Validation/TokenRequestValidator.cs]: You are sure that you dont want to log that you are removing them?

If a developer added them without knowing that they are not valid they may wonder why they are not there and there is no error.

Comment (leastprivilege) [reply src/IdentityServer4/Validation/TokenRequestValidator.cs]: feel free to do a PR - log at debug level

Comment (leastprivilege) [reply src/IdentityServer4/Validation/TokenRequestValidator.cs]: PR?

Comment (LindaLawton) [reply src/IdentityServer4/Validation/TokenRequestValidator.cs]: Probably not I have some tasks at work that need to be completed before I go on vacation.    As its your code you probably know exact where to put it anyway.

**Pull request merged**